akka-http-core/src/test/scala/akka/http/engine/client/ConnectionPoolSpec.scala

```diff
@@ -0,0 +1,358 @@
+
+      requestInSub.sendNext(HttpRequest(uri = "/a") -> 42)
+      requestInSub.sendNext(HttpRequest(uri = "/crash") -> 43)
+      responseOutSub.request(2)
+
+      val remainingResponsesToKill = new AtomicInteger(1)
+      override def mapServerSideOutboundRawBytes(bytes: ByteString): ByteString =
+        if (bytes.utf8String.contains("/crash") && remainingResponsesToKill.decrementAndGet() >= 0)
+          sys.error("CRASH BOOM BANG")
+        else bytes
+
+      val responses = Seq(responseOut.expectNext(), responseOut.expectNext())
+
+      responses mustContainLike { case (Success(x), 42) ⇒ requestUri(x) should endWith("/a") }
+      responses mustContainLike { case (Success(x), 43) ⇒ requestUri(x) should endWith("/crash") }
+    }
+
+    "respect the configured `maxRetries` value" in new TestSetup(autoAccept = true) {
+      val (requestIn, responseOut, responseOutSub, hcp) = cachedHostConnectionPool[Int](maxRetries = 4)
+
+      val requestInSub = requestIn.expectSubscription()
+      requestInSub.expectRequest(4)
+
+      requestInSub.sendNext(HttpRequest(uri = "/a") -> 42)
+      requestInSub.sendNext(HttpRequest(uri = "/crash") -> 43)
+      responseOutSub.request(2)
+
+      val remainingResponsesToKill = new AtomicInteger(5)
+      override def mapServerSideOutboundRawBytes(bytes: ByteString): ByteString =
+        if (bytes.utf8String.contains("/crash") && remainingResponsesToKill.decrementAndGet() >= 0)
+          sys.error("CRASH BOOM BANG")
+        else bytes
+
+      val responses = Seq(responseOut.expectNext(), responseOut.expectNext())
+
+      responses mustContainLike { case (Success(x), 42) ⇒ requestUri(x) should endWith("/a") }
+      responses mustContainLike { case (Failure(x), 43) ⇒ x.getMessage should include("reset by peer") }
+      remainingResponsesToKill.get() shouldEqual 0
+    }
+
+    "automatically shutdown after configured timeout periods" in new TestSetup() {
+      val (_, _, _, hcp) = cachedHostConnectionPool[Int](idleTimeout = 1.second, shutdownGracePeriod = 1.second)
+      val gateway = Await.result(hcp.gateway, 500.millis)
+      gateway.whenShuttingDown.isCompleted shouldEqual false
+      val whenShutdownCompleted = Await.result(gateway.whenShuttingDown, 1500.millis) // verify shutdown start (after idle)
+      Await.result(whenShutdownCompleted, 1500.millis) // verify shutdown completed (after grace period)
+    }
+
+    "transparently restart after idle shutdown" in new TestSetup() {
+      val (requestIn, responseOut, responseOutSub, hcp) =
+        cachedHostConnectionPool[Int](idleTimeout = 1.second, shutdownGracePeriod = 1.second)
+
+      val gateway = Await.result(hcp.gateway, 500.millis)
+      gateway.whenShuttingDown.isCompleted shouldEqual false
+      val whenShutdownCompleted = Await.result(gateway.whenShuttingDown, 1500.millis) // verify shutdown start (after idle)
+      Await.result(whenShutdownCompleted, 1500.millis) // verify shutdown completed (after grace period)
```

Comment:
This step only confirms that the shutdown is indeed completed eventually. We could also leave this out or move it to the very end. Once a shutdown has been started it will always be completed after the grace period.
Any request scheduled onto the same endpoint & pool config will end up starting a new pool instance.
The grace period only protects against not loosing racy requests.